nemo_curator/filters/heuristic_filter.py

```diff
@@ -633,3 +638,119 @@ def score_document(self, text):
 
     def keep_document(self, score):
         return score != 1
+
+
+class LengthRatioFilter(DocumentFilter):
+    """
+    For bitext cleaning.
+    If the ratio between source and target tokens is not within a specified range then discard. Either direction (src/tgt, tgt/src) is considered. See mosesdecoder/scripts/training/clean-corpus-n.perl for details
+    """
+
+    def __init__(self, max_ratio=3, src_lang="en", tgt_lang="en"):
+        super().__init__()
+        self._max_ratio = max_ratio
+        self._src_word_splitter = get_word_splitter(src_lang)
+        self._tgt_word_splitter = get_word_splitter(tgt_lang)
+        self._name = "length_ratio"
+
+    def score_document(self, bitext_tuple):
+        src_len = len(self._src_word_splitter(bitext_tuple.iloc[0].strip()))
```

Comment:
I don't like the `.iloc[0]` pattern. It seems fragile and error prone, but I see why you did it. It's the result of a poor design inherited from the `ScoreFilter` with `DocumentFilter` paradigm. I'm going to suggest you abandon the previous design for something better. I'll explain here but let me know if I need to clarify anything.

Instead of having two separate classes like `JointScoreFilter` that accepts a `DocumentFilter`, have a single base class called `BitextFilter` (or whatever name you prefer) that your other filters inherit from. The `JointFilter` definition would look something like this:

```python
class BitextFilter:
  def __init__():
    # Roughly same init as the JointScoreFilter

  def __call__():
    # Same call as the JointScoreFilter

  @abstractmethod
  def score_document():
    pass

  @abstractmethod
  def keep_document():
    pass

```

Then, all the filters like this inherit from `BitextFilter` directly. This also gives them access to the `src_field` and `tgt_field` as member variables. I know it's a bit of a change, but I'm trying to learn from the headaches I've run into with our design and make it easier for you and your users. I plan on porting the old filters to a similar design too, I just haven't gotten around to it.

Reply:
Can you clarify on why `.iloc[0]` is fragile? My exposure to `DataFrame`-like objects is probably too little to understand.

If I understand correctly, your comment basically proposes two ideas:

- unify filter objects inheriting `DocumentFilter` and filter modules like `ScoreFilter`, `JointScoreFilter`, etc.
- unify `JointScoreFilter` and `ParallelScoreFilter` into a uniform base class for bitext filtering.

As far as I understood it, I actually don't quite like either of the ideas. Let me explain why.

Basically, `JointScoreFilter` and `ParallelScoreFilter` has different logics. Let's say each sentence pair is denoted as `(s, t)`.

- What `JointScoreFilter` accomplishes is to compute score with a function `f(s, t)`, then filter according to the return value of the function.
- What `ParallelScoreFilter` accomplishes is to compute *two* scores with function `f(s)`, `g(t)` (note that `f` and `g` are not necessarily the same function). Then, we first see if `s` and `t` will be retained. If any of these are not retained, we'll have to throw the other side as well to maintain the sentence alignment of the data.

This should already tell you that `JointScoreFilter` and `ParallelScoreFilter` will have different logics in the `__call__` function. Hence unifying them will result in some awkward design.

I also like the idea of decoupling the high-level filtering logic (in `ScoreFilter`, `JointScoreFilter`, etc.) and the specific filtering functions (`f(s, t)`, `f(s)` and `g(t)` mentioned above). What this enables me is to re-use any the monolingual filtering functions implemented previously in `ParallelScoreFilter`, which is great.

Let me know if that makes sense.

Reply:
> - unify filter objects inheriting DocumentFilter and filter modules like ScoreFilter, JointScoreFilter, etc.

> ...

> What this enables me is to re-use any monolingual filtering functions implemented previously in ParallelScoreFilter, which is great.

This makes sense. I'm fine keeping the `DocumentFilter` logic separate from the `ScoreFilter` and `JointScoreFilter` modules because of this.

> - unify JointScoreFilter and ParallelScoreFilter into a uniform base class for bitext filtering.

Sorry for not being clear, this was not my intention. I agree they need to be separate since they are fundamentally two different operations.

> Can you clarify on why .iloc[0] is fragile? My exposure to DataFrame-like objects is probably too little to understand.

This is now my only concern with this filter. The fragility comment comes from the fact that it's easy to accidentally change the order of the items that get passed in here by changing the order the `fields` list gets populated in the `JointScoreFilter` class. If you could somehow have them be different arguments in the function, or identify them by unique names that would be preferable to the vague index.

Also, you mentioned that you liked the `DocumentFilter` to be separate from `ScoreFilter` since it allows you to reuse the filtering logic across methods. However, in this case it's harming you since this `DocumentFilter` method cannot be used with a `ScoreFilter`. You might want to define a new base class for this.